File: src/parser/infix_to_postfix.cpp
#include <bits/stdc++.h>
#define el "\n"
#define Hossiny ios_base::sync_with_stdio(0); cin.tie(NULL); cout.tie(NULL);

using namespace std;

// precedence levels : * > . > |
int precedence(char c) {
    if (c == '*') return 3;
    if (c == '.') return 2; 
    if (c == '|') return 1;
    return 0;
}

string insertConcatOperators(const string& regex) {
    string result;
    for (size_t i = 0; i < regex.length(); ++i) {
        char curr = regex[i];
        result += curr;

        if (i + 1 < regex.length()) {
            char next = regex[i + 1];

            bool currIsToken = isalnum(curr) || curr == '*' || curr == ')';
            bool nextIsToken = isalnum(next) || next == '(';

            if (currIsToken && nextIsToken) {
                result += '.';
            }
        }
    }
    return result;
}


string infixToPostfix(string expr) {
    stack<char> opStack;
    string postfix;

    // Replace '+' with '|' for union
    for (char& c : expr) {
        if (c == '+') c = '|';
    }

    for (size_t i = 0; i < expr.length(); ++i) {
        char c = expr[i];
        if (c == ' ') continue;

        if (isalnum(c)) {
            postfix += c;
            postfix += ' ';
        }
        else if (c == '(') {
            opStack.push(c);
        }
        else if (c == ')') {
            while (!opStack.empty() && opStack.top() != '(') {
                postfix += opStack.top();
                postfix += ' ';
                opStack.pop();
            }
            if (!opStack.empty()) opStack.pop(); // remove '('
        }
        else {
            while (!opStack.empty() && precedence(c) <= precedence(opStack.top())) {
                postfix += opStack.top();
                postfix += ' ';
                opStack.pop();
            }
            opStack.push(c);
        }
    }

    while (!opStack.empty()) {
        postfix += opStack.top();
        postfix += ' ';
        opStack.pop();
    }

    return postfix;
}


int main() {
    Hossiny

    string input;
    cout << "Enter a regex expression: ";
    getline(cin, input);

    string prepared = insertConcatOperators(input);
    string postfix = infixToPostfix(prepared);

    cout << "\nPrepared Infix (with . inserted): " << prepared << el;
    cout << "Postfix Expression: " << postfix << el;

    return 0;
}

File: src/parser/infix_to_postfix.py

import re

def validate_regex(expr):
    # Allow only: a-z A-Z 0-9 | * + ? ( ) and whitespace
    pattern = r"^[a-zA-Z0-9|*+?.() ]*$"
    if not re.fullmatch(pattern, expr):
        raise ValueError("❌ Invalid characters in regex. Allowed: a-z, A-Z, 0-9, |, *, +, ?, ., (, )")
    
def precedence(c):
    if c in ['*', '+', '?']:
        return 4
    elif c == '.':
        return 3
    elif c == '|':
        return 2
    else:
        return 0

def insert_concat_operators(regex):
    result = ""
    for i in range(len(regex)):
        curr = regex[i]
        result += curr

        if i + 1 < len(regex):
            next_char = regex[i + 1]

            # Characters that imply end of one operand
            curr_is_end = curr.isalnum() or curr in ['*', '+', '?', ')']
            # Characters that imply start of next operand
            next_is_start = next_char.isalnum() or next_char == '('

            if curr_is_end and next_is_start:
                result += '.'
    return result

def infix_to_postfix(expr):
    validate_regex(expr)
    op_stack = []
    postfix = ""

    for c in expr:
        if c == ' ':
            continue
        elif c.isalnum():
            postfix += c + ' '
        elif c == '(':
            op_stack.append(c)
        elif c == ')':
            while op_stack and op_stack[-1] != '(':
                postfix += op_stack.pop() + ' '
            if op_stack: op_stack.pop()  # remove '('
        else:
            while op_stack and precedence(c) <= precedence(op_stack[-1]):
                postfix += op_stack.pop() + ' '
            op_stack.append(c)

    while op_stack:
        postfix += op_stack.pop() + ' '

    return postfix.strip()
